File: macos/Sources/SimpleLink/TransferProgressState.swift
import Foundation

enum TransferDirection {
    case none
    case sending
    case receiving
}

struct TransferProgressState {
    var direction: TransferDirection = .none
    var done: Int64 = 0
    var total: Int64 = 0
    var label: String = ""

    var active: Bool {
        direction == .sending || direction == .receiving
    }

    var fraction: Double {
        guard total > 0 else { return 0 }
        return min(1, Double(done) / Double(total))
    }
}

File: android/app/src/main/java/com/simplelink/app/TransferProgressState.kt
package com.simplelink.app

enum class TransferDirection {
    None,
    Sending,
    Receiving,
    Queued
}

data class TransferProgressState(
    val direction: TransferDirection = TransferDirection.None,
    val done: Long = 0,
    val total: Long = 0,
    val queuedFiles: Int = 0,
    val label: String = ""
) {
    val active: Boolean
        get() = direction == TransferDirection.Sending ||
            direction == TransferDirection.Receiving

    val fraction: Float
        get() = if (total <= 0) 0f else (done.toFloat() / total.toFloat()).coerceIn(0f, 1f)
}

object TransferProgress {
    fun percent(done: Long, total: Long): Int {
        if (total <= 0) return 0
        return minOf(100, ((done * 100) / total).toInt())
    }

    fun sending(done: Long, total: Long): String =
        "${percent(done, total)}% — Sending…"

    fun receiving(done: Long, total: Long): String =
        "${percent(done, total)}% — Receiving…"

    fun queued(count: Int): String =
        if (count == 1) "1 file queued — waiting for Mac…"
        else "$count files queued — waiting for Mac…"
}
